### PR TITLE
[Feature] Add PARE

### PR DESCRIPTION
**Pull request by WYJSJTU**
Feature:

- Training and testing code for PARE: Part Attention Regressor for 3D Human Body Estimation [ICCV 2021].
- Achieving 49.35mm PA-MPJPE, 81.79 MPJPE on 3DPW, compared to the original implementation with 50.9mm PA-MPJPE, 82 MPJPE.
- Provided with detailed pre-train and training config.

### CODE REVIEW

File: tools/slurm_train_pare.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+PYTHONPATH="$(dirname $0)/..":$PYTHONPATH \
+srun -p ${PARTITION} \
+    --job-name=${JOB_NAME} \
+    --gres=gpu:${GPUS_PER_NODE} \
+    --ntasks=${GPUS} \
+    --ntasks-per-node=${GPUS_PER_NODE} \
+    --cpus-per-task=${CPUS_PER_TASK} \
+    --kill-on-bad-exit=1 \
+    ${SRUN_ARGS} \
+    python -u tools/train.py ${CONFIG} --work-dir=${WORK_DIR} --launcher="slurm" ${PY_ARGS}
```
Comment (caizhongang) [inline tools/slurm_train_pare.sh:25]: Can we use the general `slurm_train.sh`? It seems we do not need a special script just for pare.

File: tools/train.py [hunk 2]
```diff
@@ -96,7 +96,7 @@ def main():
     # create work_dir
     mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))
     # dump config
-    cfg.dump(osp.join(cfg.work_dir, osp.basename(args.config)))
+    # cfg.dump(osp.join(cfg.work_dir, osp.basename(args.config)))
```
Comment (caizhongang) [inline tools/train.py:99]: This line seems to affect other methods.

File: mmhuman3d/utils/geometry.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def rot6d_to_rotmat(x):
     Output:
         (B,3,3) Batch of corresponding rotation matrices
     """
-    x = x.view(-1, 3, 2)
+    x = x.reshape(-1, 3, 2)
```
Comment (caizhongang) [inline mmhuman3d/utils/geometry.py:63]: It seems we can refactor geoemtry to accept both numpy and tensor?

**Review by caizhongang — COMMENTED** [review]
Please also add README in configs/pare.

File: configs/smplify/smplify.py [hunk 4]
```diff
@@ -6,13 +6,13 @@
     num_betas=10,
```
[the remark this answers is not in the document]
Comment (WYJSJTU) [reply configs/smplify/smplify.py:6]: Already changed back to the original smplify.py.

File: mmhuman3d/utils/geometry.py (hunk 3 above)
Comment (WYJSJTU) [reply mmhuman3d/utils/geometry.py:63]: Add an 'if' statement to accept both numpy and tensor.

File: mmhuman3d/core/visualization/visualize_smpl.py [hunk 5]
```diff
@@ -472,6 +471,8 @@ def render_smpl(
     kp3d: Optional[Union[np.ndarray, torch.Tensor]] = None,
     mask: Optional[Union[np.ndarray, List[int]]] = None,
     vis_kp_index: bool = False,
+    # visualize render progress
+    vis_progress: bool = True,
```
[the remark this answers is not in the document]
Comment (WYJSJTU) [reply mmhuman3d/core/visualization/visualize_smpl.py:475]: Removed the comments and change the argument to 'verbose'.

File: mmhuman3d/core/visualization/renderer/torch3d_renderer/render_runner.py [hunk 6]
```diff
@@ -29,6 +29,7 @@ def render(renderer: Union[nn.Module, dict],
            batch_size: int = 5,
            return_tensor: bool = False,
            no_grad: bool = False,
+           vis_progress: bool = True,
```
[the remark this answers is not in the document]
Comment (WYJSJTU) [reply mmhuman3d/core/visualization/renderer/torch3d_renderer/render_runner.py:32]: Already changed to 'verbose'.

File: mmhuman3d/apis/train.py [hunk 7]
```diff
@@ -163,4 +161,5 @@ def train_model(model,
         runner.resume(cfg.resume_from)
     elif cfg.load_from:
         runner.load_checkpoint(cfg.load_from)
+    logger.info('start_train')
```
[the remark this answers is not in the document]
Comment (WYJSJTU) [reply mmhuman3d/apis/train.py:164]: Already deleted.

File: mmhuman3d/models/heads/pare_head.py [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.runner.base_module import BaseModule
+from torch.nn.modules.utils import _pair
+
+from mmhuman3d.utils.geometry import rot6d_to_rotmat
+from ..builder import HEADS
+
+BN_MOMENTUM = 0.1
```
Comment (caizhongang) [inline mmhuman3d/models/heads/pare_head.py:11]: Should `BN_MOMENTUM` be included in the config?

File: mmhuman3d/models/architectures/mesh_estimator.py [hunk 9]
```diff
@@ -431,19 +440,30 @@ def compute_keypoints3d_loss(self, pred_keypoints3d: torch.Tensor,
         pred_keypoints3d = pred_keypoints3d - pred_pelvis[:, None, :]
         loss = self.loss_keypoints3d(
             pred_keypoints3d, gt_keypoints3d, reduction_override='none')
-        valid_pos = keypoints3d_conf > 0
-        if keypoints3d_conf[valid_pos].numel() == 0:
-            return torch.Tensor([0]).type_as(gt_keypoints3d)
-        loss = torch.sum(loss * keypoints3d_conf)
-        loss /= keypoints3d_conf[valid_pos].numel()
+        if has_keypoints3d is None:
```
Comment (caizhongang) [inline mmhuman3d/models/architectures/mesh_estimator.py:443]: Some comments would be great. i.e. explain under what circumstances will has_keypoints3d be None?

File: mmhuman3d/models/heads/pare_head.py (hunk 8 above)
Comment (WYJSJTU) [reply mmhuman3d/models/heads/pare_head.py:11]: Change it as an argument of the Parehead with a default value of 0.1.

File: configs/pare/hrnet_w32_conv_pare_coco_cache_general.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
+use_adversarial_train = True
```
[the remark this answers is not in the document]
Comment (WYJSJTU) [reply configs/pare/hrnet_w32_conv_pare_coco_cache_general.py:1]: Changed the name to hrnet_w32_conv_pare_coco_cache and added non-cache config.

File: mmhuman3d/models/losses/crossentropy_loss.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
+import torch
```
[the remark this answers is not in the document]
Comment (WYJSJTU) [reply mmhuman3d/models/losses/crossentropy_loss.py]: Solved in commit be723f6a93bed663e8d15675046cdca22f503e4b.

File: configs/pare/hrnet_w32_conv_pare_coco.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        use_keypoint_attention=True,
+        backbone='hrnet_w32-conv',
+    ),
+    body_model_train=dict(
+        type='SMPL',
+        keypoint_src='smpl_54',
+        keypoint_dst='smpl_24',
+        model_path='data/body_models/smpl',
+        keypoint_approximate=True,
+        extra_joints_regressor='data/J_regressor_extra.npy'),
```
[the remark this answers is not in the document]
Comment (WYJSJTU) [reply configs/pare/hrnet_w32_conv_pare_coco.py]: Solved in commit e1daeeb8d9ba667fce270e2a37be1fce75f4a2d1, thanks for your review comments.

**Review by caizhongang — APPROVED** [review]

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @WYJSJTU ！First of all, we want to express our gratitude for your significant PR in the mmhuman3d project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤

**Pull request merged**